File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    answer = 0
    
    n = len(maps)
    m = len(maps[0])
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    def bfs(start_x, start_y, end_x, end_y):
        dq = deque()
        visited = [[False] * m for _ in range(n)]
        
        visited[start_x][start_y] = True
        dq.append((start_x, start_y, 0))
        
        while dq:
            x, y, cost = dq.popleft()
            if x == end_x and y == end_y:
                return cost
            
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                
                if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny] and maps[nx][ny] != 'X':
                    visited[nx][ny] = True
                    dq.append((nx, ny, cost+1))
        return -1
    
    S = L = E = None
    for i in range(n):
        for j in range(m):
            if maps[i][j] == "S":
                S = (i, j)
            elif maps[i][j] == "L":
                L = (i, j)
            elif maps[i][j] == "E":
                E = (i, j)
        if S and L and E:
            break
    
    start_to_lever = bfs(S[0], S[1], L[0], L[1])
    lever_to_end = bfs(L[0], L[1], E[0], E[1])
    
    if start_to_lever == -1 or lever_to_end == -1:
        answer = -1
    else:
        answer = start_to_lever + lever_to_end
    
    return answer

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    int n, m;
    int[][] directions = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} };
    
    public int solution(String[] maps) {
        int answer = 0;
        
        n = maps.length;
        m = maps[0].length();
        
        int[] S = null, L = null, E = null;
        
        outer:
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                char cell = maps[i].charAt(j);
                if (cell == 'S') S = new int[]{i, j};
                else if (cell == 'L') L = new int[]{i, j};
                else if (cell == 'E') E = new int[]{i, j};
                
                if (S != null && L != null && E != null) {
                    break outer;
                }
            }
        }
        
        int toLever = bfs(S, L, maps);
        int toEnd = bfs(L, E, maps);
        
        if (toLever == -1 || toEnd == -1) {
            answer = -1;
        } else {
            answer = toLever + toEnd;
        }
        
        return answer;
    }
    
    int bfs(int[] start, int[] end, String[] maps) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{ start[0], start[1], 0 });
        
        boolean[][] visited = new boolean[n][m];
        visited[start[0]][start[1]] = true;
        
        while (!queue.isEmpty()) {
            int[] temp = queue.poll();
            int x = temp[0], y = temp[1], cost = temp[2];
            
            if (x == end[0] && y == end[1]) return cost;
            
            for (int[] dir: directions) {
                int nx = x + dir[0], ny = y + dir[1];
                
                if (nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny]) {
                    char curr = maps[nx].charAt(ny);
                    if (curr == 'X') continue;
                    visited[nx][ny] = true;
                    queue.offer(new int[]{ nx, ny, cost + 1 });
                }
            }
        }
        
        return -1;
    }
}